### ISSUE TITLE
"Illegal constructor" on DatabaseFactory.open

### ISSUE THREAD

**Issue by retorquere** [MockingMagician/promised-db#9]
**Describe the bug**
I'm getting an "Illegal constructor" error

**To Reproduce**

```
await DatabaseFactory.open(this.name, this.version, [{
        version: this.version,
        migration: async ({ db, dbOldVersion, dbNewVersion }) => { // eslint-disable-line @typescript-eslint/require-await
          if (typeof dbNewVersion !== 'number') {
            log.info(`cache: creating ${dbNewVersion}`)
          }
          else {
            log.info(`cache: upgrading ${dbOldVersion} => ${dbNewVersion}`)
          }
          for (const store of db.objectStoreNames) {
            db.deleteObjectStore(store)
          }

          db.createObjectStore('ZoteroSerialized', { keyPath: 'itemID' })
          db.createObjectStore('touched')
          db.createObjectStore('metadata', { keyPath: 'key' })

          const context = db.createObjectStore('ExportContext', { keyPath: 'id', autoIncrement: true })
          context.createIndex('context', 'context', { unique: true })

          const stores = [
            db.createObjectStore('BetterBibTeX', { keyPath: [ 'context', 'itemID' ]}),
            db.createObjectStore('BetterBibLaTeX', { keyPath: [ 'context', 'itemID' ]}),
            db.createObjectStore('BetterCSLJSON', { keyPath: [ 'context', 'itemID' ]}),
            db.createObjectStore('BetterCSLYAML', { keyPath: [ 'context', 'itemID' ]}),
          ]
          for (const store of stores) {
            store.createIndex('context', 'context')
            store.createIndex('itemID', 'itemID')
            store.createIndex('context-itemID', [ 'context', 'itemID' ], { unique: true })
          }
        },
      }])
```

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser: Zotero
 - Version: 7.0.6

**Comment by MockingMagician** [comment]
Hello @retorquere

Your code seemed perfectly valid to me, in doubt I ran it in a real web context, and the result shows that there is no error.

Could you extend the context of your error? The whole code in which this little piece of code is executed, because I see references such as `this.version` so I imagine it lives in a class context somewhere..., as well as the whole of your stack trace?

**Comment by retorquere** [comment]
```
const Cache = new class $Cache {
  public name = 'BetterBibTeXCache'
  public version = 10

  private db: Database

  private async $open(action: string): Promise<Database> {
    try {
      return await DatabaseFactory.open(this.name, this.version, [{
        version: this.version,
        migration: async ({ db, dbOldVersion, dbNewVersion }) => { // eslint-disable-line @typescript-eslint/require-await
          if (typeof dbNewVersion !== 'number') {
            log.info(`cache: creating ${dbNewVersion}`)
          }
          else {
            log.info(`cache: upgrading ${dbOldVersion} => ${dbNewVersion}`)
          }
          for (const store of db.objectStoreNames) {
            db.deleteObjectStore(store)
          }

          db.createObjectStore('ZoteroSerialized', { keyPath: 'itemID' })
          db.createObjectStore('touched')
          db.createObjectStore('metadata', { keyPath: 'key' })

          const context = db.createObjectStore('ExportContext', { keyPath: 'id', autoIncrement: true })
          context.createIndex('context', 'context', { unique: true })

          const stores = [
            db.createObjectStore('BetterBibTeX', { keyPath: [ 'context', 'itemID' ]}),
            db.createObjectStore('BetterBibLaTeX', { keyPath: [ 'context', 'itemID' ]}),
            db.createObjectStore('BetterCSLJSON', { keyPath: [ 'context', 'itemID' ]}),
            db.createObjectStore('BetterCSLYAML', { keyPath: [ 'context', 'itemID' ]}),
          ]
          for (const store of stores) {
            store.createIndex('context', 'context')
            store.createIndex('itemID', 'itemID')
            store.createIndex('context-itemID', [ 'context', 'itemID' ], { unique: true })
          }
        },
      }])
    }
    catch (err) {
      log.error(`cache: ${action} ${this.version} failed: ${err.message}\n${err.stack}`)
      return null
    }
  }

  public async open(lastUpdated?: string): Promise<void> {
    if (this.db) throw new Error('database reopened')

    this.db = await this.$open('open')
    if (!this.db) {
      log.info('cache: could not open, delete and reopen') // #2995, downgrade 6 => 7
      await DatabaseFactory.deleteDatabase(this.name)
      this.db = await this.$open('reopen')
    }

    if (this.db && !worker) {
      const del = 'extensions.zotero.translators.better-bibtex.cache.delete'
      const metadata = { Zotero: '', BetterBibTeX: '', lastUpdated: '', ...(await this.metadata()) }
      const reasons = [
        { reason: `Zotero version changed from ${metadata.Zotero || 'none'} to ${Zotero.version}`, test: metadata.Zotero && metadata.Zotero !== metadata.Zotero },
        { reason: `Better BibTeX version changed from ${metadata.BetterBibTeX || 'none'} to ${version}`, test: metadata.BetterBibTeX && metadata.BetterBibTeX !== version },
        { reason: `cache gap found ${metadata.lastUpdated} => ${lastUpdated}`, test: (lastUpdated || false) && metadata.lastUpdated && lastUpdated !== metadata.lastUpdated },
        { reason: 'marked for deletion', test: Zotero.Prefs.get(del) || false },
      ]
      const reason = reasons.filter(r => r.test).map(r => r.reason).join(' and ') || false
      Zotero.Prefs.clear(del)

      log.info(`cache: ${JSON.stringify(metadata)} + ${JSON.stringify({ Zotero: Zotero.version, BetterBibTeX: version, lastUpdated })} => ${JSON.stringify(reasons)} => ${reason}`)
      if (reason) {
        log.info(`cache: reopening because ${reason}`)
        this.db.close()
        await DatabaseFactory.deleteDatabase(this.name)
        this.db = await this.$open('upgrade')
      }
      const tx = this.db.transaction('metadata', 'readwrite')
      const store = tx.objectStore('metadata')
      await store.put({ key: 'Zotero', value: Zotero.version })
      await store.put({ key: 'BetterBibTeX', value: version })
      await tx.commit()
    }
  }

  public close(): void {
    this.db?.close()
    this.db = null
  }
}

await Cache.open()
```

**Comment by MockingMagician** [comment]
Two questions:

- Can you provide me with the complete stack-trace of the error?
- Can you tell me if your code is compiled before being used, and if so, with what tools and process?

**Comment by retorquere** [comment]
> Can you provide me with the complete stack-trace of the error?

Ah sorry, yes; my environment doesn't support sourcemaps, so I've had to translate by hand:

node_modules/@idxdb/promised/dist/src/database-factory.js: 13
node_modules/@idxdb/promised/dist/src/database-factory.js: 40
called by "return await DatabaseFactory.open"

> Can you tell me if your code is compiled before being used, and if so, with what tools and process?

It is indeed, compiled using `esbuild`. The $open function compiles to

```
async $open(action) {
          try {
            return await DatabaseFactory.open(this.name, this.version, [{
              version: this.version,
              migration: async ({ db, dbOldVersion, dbNewVersion }) => {
                if (typeof dbNewVersion !== "number") {
                  log.info(`cache: creating ${dbNewVersion}`);
                } else {
                  log.info(`cache: upgrading ${dbOldVersion} => ${dbNewVersion}`);
                }
                for (const store of db.objectStoreNames) {
                  db.deleteObjectStore(store);
                }
                db.createObjectStore("ZoteroSerialized", { keyPath: "itemID" });
                db.createObjectStore("touched");
                db.createObjectStore("metadata", { keyPath: "key" });
                const context = db.createObjectStore("ExportContext", { keyPath: "id", autoIncrement: true });
                context.createIndex("context", "context", { unique: true });
                const stores = [
                  db.createObjectStore("BetterBibTeX", { keyPath: ["context", "itemID"] }),
                  db.createObjectStore("BetterBibLaTeX", { keyPath: ["context", "itemID"] }),
                  db.createObjectStore("BetterCSLJSON", { keyPath: ["context", "itemID"] }),
                  db.createObjectStore("BetterCSLYAML", { keyPath: ["context", "itemID"] })
                ];
                for (const store of stores) {
                  store.createIndex("context", "context");
                  store.createIndex("itemID", "itemID");
                  store.createIndex("context-itemID", ["context", "itemID"], { unique: true });
                }
              }
            }]);
          } catch (err) {
            log.error(`cache: ${action} ${this.version} failed: ${err.message}
${err.stack}`);
            return null;
          }
        }
```

but I'm willing to accept that Zotero doesn't support this library. The Zotero XPCOM environment is a weird mix between  mostly browser-ish and a dash of node-ish.

Ah but I think that explains it. `indexedDB` is defined in this environment, `IDBFactory` is not. But [MDN](https://developer.mozilla.org/en-US/docs/Web/API/IDBFactory) says

> The object that implements the interface is window.indexedDB. You open — that is, create and access — and delete a database with this object, and not directly with IDBFactory.

**Comment by MockingMagician** [comment]
Okay, from the stack trace it's much clearer.

My first question is. Why don't you fall within the context of line 11? You fall into line 13, which is a context outside the web-worker, outside the browser, since self isn't defined... 

I wonder if adapting my factory and making it more permissive and broader in terms of context wouldn't solve your problem...

**Comment by retorquere** [comment]
Ah because `self` is undefined in my environment, but just plain `indexedDB` is defined. Why not test for `typeof indexedDB`? Is there a benefit to going through `self`?

**Comment by retorquere** [comment]
But does using IDBFactory actually work? If I run

```
var _factory =  new IDBFactory()
```

in my Edge console, I get the same `Illegal constructor` error. It's not just Zotero.

**Comment by MockingMagician** [comment]
> Ah because `self` is undefined in my environment, but just plain `indexedDB` is defined. Why not test for `typeof indexedDB`? Is there a benefit to going through `self`?

This is simply good practice. `self` represents the global context for both (browser or webworker). This theoretically avoids multiple context checks and allows direct access to the API regardless of context.

But given your problem, it seems that this isn't always the case for all contexts.

I'll transform the factory so that it can handle these special cases.

---

> But where does using IDBFactory actually work? If I run
> 
> ```
> var _factory =  new IDBFactory()
> ```
> 
> in my Edge console, I get the same `Illegal constructor` error. It's not just Zotero.

`new IDBFactory()` is just a node.js implementation of the indexedDB API. This allows me to perform all unit tests in node without using a browser.

In short, you should never run into this unless you're running unit tests from the repository.

That's also why I've started implementing tests to run in all browsers, to make sure the API works equally well everywhere... And that it's really `self.indexedDB` that's being called

**Comment by retorquere** [comment]
Wouldn't

```
(typeof process !== 'undefined') && (process.release.name === 'node')
```

give you a positive test of running under node rather than taking it to be the fallback position?

But when I run node, I don't have IDBFactory:

```
% node
Welcome to Node.js v22.8.0.
Type ".help" for more information.
> IDBFactory
Uncaught ReferenceError: IDBFactory is not defined
```

so maybe IDBFactory comes from a library?

**Comment by retorquere** [comment]
What library is providing `IDBFactory`? I wonder why it is not providing `indexedDB`.

**Comment by MockingMagician** [comment]
> so maybe IDBFactory comes from a library?

Yeap, this library is loaded during unit test, takes a look at `component-test.initializer.ts`

---

changes done and avalilable in v4.0.3

**Comment by MockingMagician** [comment]
> Wouldn't
> 
> ```
> (typeof process !== 'undefined') && (process.release.name === 'node')
> ```
> 
> give you a positive test of running under node rather than taking it to be the fallback position?

I agree, the fallback solution must not be present, it's a very bad conception.

I'd find a cleaner solution for the unit-testing part, which runs on the fake IndexedDB library.

**Comment by retorquere** [comment]
> > What library is providing `IDBFactory`? I wonder why it is not providing `indexedDB`.
> 
> fake-indexeddb

but that also provides `indexedDB`.

**Comment by MockingMagician** [comment]
indeed, I feel a bit stupid not to have seen it :sweat_smile:

**Comment by MockingMagician** [comment]
Fully removed...

**Comment by MockingMagician** [comment]
@retorquere  let me know if this solved your problem
